sh_scrapy/settings.py

```diff
@@ -95,12 +101,21 @@ def _load_addons(addons, settings, merged_settings, priority=0):
             raise
         skey = _get_component_base(settings, addon['type'])
         components = settings[skey]
-        path = update_classpath(addon_path)
-        components[path] = addon['order']
+        _update_component_order(components, addon_path, addon['order'])
         merged_settings.set(skey, components)
         merged_settings.setdict(addon['default_settings'], priority)
 
 
+def _merge_with_keeping_order(settings, updates):
+    for setting_key, value in updates.items():
+        if not isinstance(value, dict):
+            settings.set(setting_key, value, priority='cmdline')
+            continue
+        components = settings[setting_key]
+        for path, order in value.items():
+            _update_component_order(components, path, order)
```

Comment:
@vshlapakov have you ever noticed that dicts or settings merged here are actually the same objects and when you update say `EXTENSIONS_BASE` in merged_settings you're actually mutating `EXTENSIONS_BASE` in original settings?

the branch

```
-> _merge_with_keeping_order(settings, merged_settings.copy_to_dict())
(Pdb) settings['EXTENSIONS_BASE']
<scrapy.settings.BaseSettings object at 0x107e3bc10>
(Pdb) merged_settings['EXTENSIONS_BASE']
<scrapy.settings.BaseSettings object at 0x107e3bc10>
```

v0.6.0
```
-> s.setdict(o, priority='cmdline')
(Pdb) !s['EXTENSIONS_BASE']
<scrapy.settings.BaseSettings object at 0x102a8b550>
(Pdb) !o['EXTENSIONS_BASE']
<scrapy.settings.BaseSettings object at 0x102a8b550>
```

v0.6.0 and Scrapy 1.0.5
```
-> s.setdict(o, priority='cmdline')
(Pdb) !s['EXTENSIONS_BASE']
{'scrapy.extensions.feedexport.FeedExporter': 0, 'scrapy.extensions.memdebug.MemoryDebugger': 0, 'scrapy.extensions.memusage.MemoryUsage': 0, 'scrapy.extensions.logstats.LogStats': 0, 'scrapy.extensions.debug.StackTraceDump': 0, 'scrapy.telnet.TelnetConsole': 0, 'scrapy.extensions.corestats.CoreStats': 0, 'scrapy.extensions.closespider.CloseSpider': 0, 'sh_scrapy.extension.HubstorageExtension': 100, 'scrapy.extensions.spiderstate.SpiderState': 0, 'scrapy.extensions.throttle.AutoThrottle': 0}
(Pdb) !id(s['EXTENSIONS_BASE'])
4451296712
(Pdb) !id(o['EXTENSIONS_BASE'])
4451296712
```